deployable/java/src/main/java/envtest/deployable/DeployableApplication.java

```diff
@@ -0,0 +1,157 @@
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpExchange;
+
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.Map;
+import java.lang.reflect.Method;
+
+import java.io.BufferedReader;
+import java.net.URL;
+import java.io.InputStreamReader;
+import java.net.MalformedURLException;
+import java.net.HttpURLConnection;
+
+/**
+ * This class serves as an entry point for the Spring Boot app
+ * Here, we check to ensure all required environment variables are set
+ */
+@SpringBootApplication
+public class DeployableApplication {
+
+    private static final org.slf4j.Logger logger = LoggerFactory.getLogger(DeployableApplication.class);
+
+    private static String getProjectId() throws RuntimeException {
+        try {
+            // try reading from service account
+            GoogleCredentials credentials = GoogleCredentials.getApplicationDefault();
+            if (credentials instanceof ServiceAccountCredentials) {
+              String id = ((ServiceAccountCredentials) credentials).getProjectId();
+              if (id != null){
+                  return id;
+              }
+            }
+            // try grabbing from metadata server
+            URL url = new URL("http://metadata.google.internal/computeMetadata/v1/project/project-id");
+            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+            conn.setRequestProperty("Metadata-Flavor", "Google");
+            BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream(), "UTF-8"));
+            String line = reader.readLine();
+            return line;
+        } catch (IOException e){
+            System.out.println(e);
+        }
+        throw new RuntimeException("could not find project ID");
+    }
+
+    private static void startPubsubSubscription() throws IOException, RuntimeException {
+        // create variables
+        String projectId = getProjectId();
+        String topicId = System.getenv().getOrDefault("PUBSUB_TOPIC", "logging-test");
+        String subscriptionId = topicId + "-subscriber";
+        TopicName topicName = TopicName.of(projectId, topicId);
+        ProjectSubscriptionName subscriptionName = ProjectSubscriptionName.of(projectId, subscriptionId);
+        // create subscription
+        SubscriptionAdminClient subscriptionClient = SubscriptionAdminClient.create();
+        subscriptionClient.createSubscription(
+               subscriptionName,
+               topicName,
+               PushConfig.newBuilder().build(),
+               20);
+        // define callback
+        MessageReceiver receiver = (PubsubMessage message, AckReplyConsumer consumer) -> {
+          consumer.ack();
+          String fnName = message.getData().toStringUtf8();
+          Map<String, String> args = message.getAttributes();
+          triggerSnippet(fnName, args);
```

Comment:
This is hacky and outright dangerous. Instead of using reflection to dynamically invoke methods, I suggest we do it "old school" by checking supported messages and triggering the one we want based on message value.

Reply:
Too dangerous to use even for test scaffolding? My impression is that the code here will only allow execution of functions from the "Snippets" class that fit the expected signature, which seems to be what we want. Are there other security issues I'm overlooking?

This is an important point, because we've talking about generalizing this code to potentially support other use cases, and requiring hard-coded function calls rather than dynamic ones would be a huge roadblock to that.